server/log_rpc_server.go

```diff
@@ -130,13 +137,56 @@ func (t *TrillianLogRPCServer) QueueLeaves(ctx context.Context, req *trillian.Qu
 
 // AddSequencedLeaf submits one sequenced leaf to the storage.
 func (t *TrillianLogRPCServer) AddSequencedLeaf(ctx context.Context, req *trillian.AddSequencedLeafRequest) (*trillian.AddSequencedLeafResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "AddSequencedLeaf is not implemented")
+	if err := validateAddSequencedLeafRequest(req); err != nil {
+		return nil, err
+	}
+
+	batchReq := &trillian.AddSequencedLeavesRequest{
+		LogId:  req.LogId,
+		Leaves: []*trillian.LogLeaf{req.Leaf},
+	}
+	rsp, err := t.AddSequencedLeaves(ctx, batchReq)
+	if err != nil {
+		return nil, err
+	}
+	if rsp == nil {
+		return nil, status.Errorf(codes.Internal, "missing response")
+	}
+	if ln := len(rsp.Results); ln != 1 {
+		return nil, status.Errorf(codes.Internal, "unexpected count of sequenced leaves %d", ln)
+	}
+	return &trillian.AddSequencedLeafResponse{Result: rsp.Results[0]}, nil
 }
 
 // AddSequencedLeaves submits a batch of sequenced leaves to a pre-ordered log
 // for later integration into its underlying tree.
 func (t *TrillianLogRPCServer) AddSequencedLeaves(ctx context.Context, req *trillian.AddSequencedLeavesRequest) (*trillian.AddSequencedLeavesResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "AddSequencedLeaves is not implemented")
+	if err := validateAddSequencedLeavesRequest(req); err != nil {
+		return nil, err
+	}
+
+	tree, hasher, err := t.getTreeAndHasher(ctx, req.LogId, false /* readonly */)
+	if err != nil {
+		return nil, err
+	}
+	if want := trillian.TreeType_PREORDERED_LOG; tree.TreeType != want {
+		return nil, status.Errorf(codes.FailedPrecondition, "adding sequenced leaves available only for %v trees", want)
```

Comment:
Error is a bit wordy and consider adding what state the tree is in to help the caller.